HAL/keymaster/4.1/JavacardKeymaster4Device.cpp

```diff
@@ -1071,78 +1025,161 @@ Return<void> JavacardKeymaster4Device::begin(KeyPurpose purpose, const hidl_vec<
+                                   "data, status: "
+                                   << (int32_t)errorCode;
                     } else {
                         /* Store the operationInfo */
-                        oprCtx_->setOperationInfo(operationHandle, purpose, param.f.algorithm, inParams);
+                        oprCtx_->setOperationInfo(operationHandle, purpose,
+                                                  param.f.algorithm, inParams);
                     }
                 }
             }
         } else {
-            LOG(ERROR) << "INS_BEGIN_OPERATION_CMD couldn't find algorithm tag: " << (int32_t)Tag::ALGORITHM;
+            LOG(ERROR) << "INS_BEGIN_OPERATION_CMD couldn't find algorithm tag: "
+                       << (int32_t)Tag::ALGORITHM;
         }
     } else {
-        LOG(ERROR) << "INS_BEGIN_OPERATION_CMD error in getKeyCharacteristics status: " << (int32_t) errorCode;
+        LOG(ERROR)
+                << "INS_BEGIN_OPERATION_CMD error in getKeyCharacteristics status: "
+                << (int32_t)errorCode;
+    }
+    return errorCode;
+}
+
+ErrorCode JavacardKeymaster4Device::handleBeginOperation(
+    KeyPurpose purpose, const hidl_vec<uint8_t>& keyBlob,
+    const hidl_vec<KeyParameter>& inParams, const HardwareAuthToken& authToken,
+    hidl_vec<KeyParameter>& outParams, uint64_t& operationHandle,
+    OperationType& operType) {
+    ErrorCode errorCode = ErrorCode::UNKNOWN_ERROR;
+    if (operType == OperationType::PUBLIC_OPERATION) {
+        errorCode = handleBeginPublicKeyOperation(purpose, keyBlob, inParams,
+                                                  outParams, operationHandle);
+
+        // For Symmetric operations handleBeginPublicKeyOperation function
+        // returns INCOMPATIBLE_ALGORITHM error. Based on this error
+        // condition it fallbacks to private key operation.
+        if (errorCode == ErrorCode::INCOMPATIBLE_ALGORITHM) {
+            operType = OperationType::PRIVATE_OPERATION;
+        }
+    }
+
+    if (operType == OperationType::PRIVATE_OPERATION) {
+        errorCode = handleBeginPrivateKeyOperation(
+            purpose, keyBlob, inParams, authToken, outParams, operationHandle);
+    }
+
+    // check if operation handle exists.
+    if (ErrorCode::OK == errorCode && isOperationHandleExists(operationHandle)) {
+        LOG(DEBUG) << "Operation handle " << operationHandle << "already exists"
+            "in the opertion table. so aborting this opertaion.";
+        // abort the operation.
+        errorCode = abortOperation(
+            operationHandle, (operType == OperationType::PRIVATE_OPERATION));
     }
-    /* Create a new operation handle and add a entry inside the operation table map with
-     * key - new operation handle
-     * value - hal generated operation handle.
+    return errorCode;
+}
+
+Return<void> JavacardKeymaster4Device::begin(KeyPurpose purpose,
```

Comment:
@BKSSMVenkateswarlu  can we move this code up before newly added functions so that in the diff, it will show next to ::begin function.

Reply:
Modified the code as per the review comments.